utils/s3/client.cc

```diff
@@ -42,10 +45,77 @@ future<> ignore_reply(const http::reply& rep, input_stream<char>&& in_) {
     co_await util::skip_entire_stream(in);
 }
 
+class dns_connection_factory : public http::experimental::connection_factory {
+protected:
+    std::string _host;
+    int _port;
+    struct state {
+        bool initialized = false;
+        socket_address addr;
+        ::shared_ptr<tls::certificate_credentials> creds;
+        shared_future<> done;
+
+        state(future<> f) : done(std::move(f)) {}
+    };
+    lw_shared_ptr<state> _state;
+
+    void initialize(std::optional<endpoint_config::https_config> https) {
+        promise<> initialized;
+        _state = make_lw_shared<state>(initialized.get_future());
+        when_all(
+            net::dns::get_host_by_name(_host, net::inet_address::family::INET).then([port = _port] (::net::hostent&& h) {
+                return make_ready_future<socket_address>(socket_address(h.addr_list.front(), port));
+            }),
+            [https = std::move(https)] () -> future<::shared_ptr<tls::certificate_credentials>> {
+                if (!https) {
+                    auto nil = ::shared_ptr<tls::certificate_credentials>();
+                    return make_ready_future<::shared_ptr<tls::certificate_credentials>>(nil);
+                }
+
+                return do_with(tls::credentials_builder(), *https, [] (auto& cbuild, auto& https) {
+                    auto f = https.trust_file == "system" ? cbuild.set_system_trust() : cbuild.set_x509_trust_file(https.trust_file, tls::x509_crt_format::PEM);
+                    return std::move(f).then([&cbuild] {
+                        auto creds = cbuild.build_certificate_credentials();
+                        return make_ready_future<::shared_ptr<tls::certificate_credentials>>(std::move(creds));
+                    });
+                });
+            }()
+        ).then([state = _state] (auto res_tuple) {
+            state->addr = std::get<0>(res_tuple).get0();
+            state->creds = std::get<1>(res_tuple).get0();
+            state->initialized = true;
+            s3l.debug("Initialized factory, address={} tls={}", state->addr, state->creds == nullptr ? "no" : "yes");
+        }).forward_to(std::move(initialized));
+    }
```

Comment:
This code looks like it would be repeated for non-s3 use cases. So before we move the http client out of experimental we should enhance it so it's easier.